### ISSUE TITLE
Add Preprocessor option for MSVC in CMakeLists.txt

### ISSUE THREAD

**Issue by Lubart** [Brewtarget/brewtarget#746]
Can preprocessor option (Enable preprocessor conformance mode) be added to CMakeLists.txt.
That option is:

if (MSVC)
  add_compile_options(/Zc:preprocessor)
endif()

In order to have it already in VC project options.
https://learn.microsoft.com/en-us/cpp/build/reference/zc-preprocessor?view=msvc-170

**Comment by matty0ung** [comment]
Sorry you didn't get a reply on this before.  The main reason we don't support MSVC out of the box is that I haven't worked out how to script all the installation etc.  At some point, I would love to get the build working on MSVC so that we're compiling on all three mainstream C++ compilers, but, realistically, it's going to be a while before that gets to the top of my to-do list.  In the meantime, I'm happy to look at PRs if you or anyone else want to take a shot at doing this.

**Comment by danielm5** [comment]
I'd be interested on make it build with MSVC. Is there anyone else working on this?

**Comment by matty0ung** [comment]
> I'd be interested on make it build with MSVC. Is there anyone else working on this?

AFAIK no-one is currently working on it.  If you want to get either the CMake or the Meson build (or both) working on MSVC that would be great.

**Comment by danielm5** [comment]
Great! I'll do some preliminary work and post back questions/comments or just an update. Should I create a separate issue to track this work?

**Comment by matty0ung** [comment]
Cool.  I reckon we might as well keep it in this issue for now.

**Comment by danielm5** [comment]
My first question is about dependencies. I read your comment about vcpkg and conan. You've noted that conan could be an alternative but that xalan-c was missing. I did a bit of searching and I learned that xalan-c was abandoned and removed from vcpkg by its maintainer ( https://github.com/microsoft/vcpkg/pull/27401 ). Besides currently being unmaintained, it appears to have some security issues based on what I've read. Have you ever considered switching to libxml2+libxslt ?

I know this kind of a side-quest from the original task, but it could be something worth pursuing.

**Comment by matty0ung** [comment]
It's a good question.  I started looking around a while back and there seemed conflicting info about the long-term future of Xalan-C & Xerces-C.  AFAICT they are currently missing active maintainers for the latter, but it wasn't clear if that was expected to be temporary or permanent.

I hadn't discovered libxml2/libxslt, so those would be good to look at.  I haven't looked at the detail, but I'd guess that switching to them will likely be a medium-sized task, depending on how similar in conception the APIs are to Xalan & Xerces.

**Comment by matty0ung** [comment]
If you want to have a stab at getting libxml2+libxslt working, that would be cool.  Otherwise, if we want to kick the can down the road, you can, worst case, compile xalan-c from source...

**Comment by matty0ung** [comment]
Looks like https://github.com/libxmlplusplus/libxmlplusplus could be a useful C++ wrapper around libxml2.

**Comment by danielm5** [comment]
Yes, I think switching to lbxml2+libxslt could bring some benefits because they are more widely available and maintained. I'll begin looking into this to try to understand how much work it could be.

**Comment by danielm5** [comment]
Small update: I think I can adapt the code for libxml2+libxslt, however, I've gotten stuck because the mingw64 cmake debug build is broken. I've spent more than two days trying to understand the problem and trying different fixes which didn't work. Basically, no matter what I do, I get undefined reference linker errors to the symbols defined as extern, for instance Measurement::Units::milligrams and many others. Take a look at this build: https://github.com/danielm5/brewtarget/actions/runs/17713448468/job/50335197181

I gave up trying to fix this for now. I'll try to find some workaround to continue.

**Comment by danielm5** [comment]
I've unblocked myself by commenting out everything related to BeerJson. I'll try to make some progress on libxml2+libxslt and I'll come back to review that debug build issue later.
